### PR TITLE
Implement TSMT leaf insertion into advice map

### LINKED ISSUE #1041: SMT sequential inserts fail
I'm having some problems executing multiple inserts in a row against a SMT.  I receive an error of the following nature:

```
---- collections::smt::tsmt_insert_16 stdout ----
thread 'collections::smt::tsmt_insert_16' panicked at 'called `Result::unwrap()` on an `Err` value: AdviceKeyNotFound([BaseElement(15556198973653036125), BaseElement(2569263626647358186), BaseElement(12115666839279463362), BaseElement(5570871396401658231)])', test-utils/src/lib.rs:289:36
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- collections::smt::tsmt_insert_32 stdout ----
thread 'collections::smt::tsmt_insert_32' panicked at 'called `Result::unwrap()` on an `Err` value: AdviceKeyNotFound([BaseElement(6914814850777475923), BaseElement(1672658896067019021), BaseElement(11491567203765900427), BaseElement(8641048661315345931)])', test-utils/src/lib.rs:289:36

---- collections::smt::tsmt_insert_48 stdout ----
thread 'collections::smt::tsmt_insert_48' panicked at 'called `Result::unwrap()` on an `Err` value: AdviceKeyNotFound([BaseElement(6110365337961474275), BaseElement(16137260194458567299), BaseElement(3850080979430617408), BaseElement(7925048865637296829)])', test-utils/src/lib.rs:289:36
```

To recreate, replace the test brace [here](https://github.com/0xPolygonMiden/miden-vm/blob/57110fa60c27e64cfdf3c8a6fd8cb4a39a79baf2/stdlib/tests/collections/smt.rs#L204-L218) with:

```rust
fn assert_insert(
    init_smt: &TieredSmt,
    key: RpoDigest,
    old_value: Word,
    new_value: Word,
    new_root: RpoDigest,
) {
    let old_root = init_smt.root();
    let source = r#"
        use.std::collections::smt

        begin
            dupw.1 movdnw.3
            exec.smt::insert

            dropw swapw
            push.1.0.0.1

            exec.smt::insert
        end
    "#;
```

Apologies for lack of a better example.

### PR DESCRIPTION
**Pull request by bobbinth**
This PR introduces new advice injector (`adv.insert_hperm`) which hashes the values at the top of the stack and uses the result to insert a new entry into the advice map. This advice injector is then used in the TSMT insert procedures to insert pre-images of newly created leaves into the advice map.

Fixes #1041.

### CODE REVIEW

File: stdlib/tests/collections/smt.rs [hunk 1]
```diff
@@ -269,61 +337,37 @@ fn assert_insert(
...
+    let mut new_adv_map = process.advice_provider.map().clone();
+    for (key, value) in adv_map.iter() {
+        let old_value = new_adv_map.remove(key).unwrap();
+        assert_eq!(value, &old_value);
+    }
+
+    // make sure the remaining values in the advice map are the same as expected new entries
+    assert_eq!(new_adv_map.len(), new_map_entries.len());
+    for (key, val) in new_map_entries {
+        let old_val = new_adv_map.get(key).unwrap();
```
Comment (Al-Kindi-0) [inline stdlib/tests/collections/smt.rs:360]: minor nit: I wonder whether or not `old_val` is appropriate as a name for this variable

**Review by Al-Kindi-0 — APPROVED** [review]
Looks great, thank you @bobbinth !
Just a minor nit from my side.

**Review by frisitano — APPROVED** [review]
Looks great, thanks! I've tested this downstream and it's working as expected.  I left a minor remark inline regarding the advice injector documentation.

File: core/src/operations/decorators/advice.rs [hunk 2]
```diff
@@ -214,6 +214,21 @@ pub enum AdviceInjector {
...
+
+    /// Reads three words from the operand stack and inserts the top two words into the advice map
+    /// under the key defined by applying an RPO permutation to all three words.
+    ///
+    /// Inputs:
+    ///   Operand stack: [B, A, C, ...]
+    ///   Advice map: {...}
+    ///
+    /// Outputs:
+    ///   Operand stack: [B, A, ...]
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply core/src/operations/decorators/advice.rs:226]: Fixed!

File: processor/src/decorators/adv_map_injectors.rs [hunk 3]
```diff
@@ -72,6 +76,50 @@ where
...
 
+    /// Reads three words from the operand stack and inserts the top two words into the advice map
+    /// under the key defined by applying an RPO permutation to all three words.
+    ///
+    /// Inputs:
+    ///   Operand stack: [B, A, C, ...]
+    ///   Advice map: {...}
+    ///
+    /// Outputs:
+    ///   Operand stack: [B, A, ...]
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply processor/src/decorators/adv_map_injectors.rs:87]: Fixed!

File: stdlib/tests/collections/smt.rs (hunk 1 above)
Comment (bobbinth) [reply stdlib/tests/collections/smt.rs:360]: Yeah - I've replaced it with `init_value`.

**Pull request merged**